test/boost/commitlog_test.cc

```diff
@@ -545,6 +551,46 @@ SEASTAR_TEST_CASE(test_commitlog_reader_produce_exception){
         });
 }
 
+SEASTAR_TEST_CASE(test_commitlog_segment_truncation){
+    commitlog::config cfg;
+    cfg.commitlog_segment_size_in_mb = 1;
+    return cl_test(cfg, [](commitlog& log) {
+        auto rps = make_lw_shared<std::vector<db::replay_position>>();
+        return do_until([rps]() {return rps->size() > 1;},
+                    [&log, rps]() {
+                        auto uuid = utils::UUID_gen::get_time_UUID();
+                        sstring tmp = "hej bubba cow";
+                        return log.add_mutation(uuid, tmp.size(), db::commitlog::force_sync::no, [tmp](db::commitlog::output& dst) {
+                                    dst.write(tmp.data(), tmp.size());
+                                }).then([&log, rps](rp_handle h) {
+                                    BOOST_CHECK_NE(h.rp(), db::replay_position());
+                                    rps->push_back(h.release());
+                                });
+                    }).then([&log]() {
+                        return log.sync_all_segments();
+                    }).then([&log, rps] {
+                        auto segments = log.get_active_segment_names();
+                        BOOST_REQUIRE(!segments.empty());
+                        auto seg = segments[0];
+                        // must truncate past header size, otherwise entry is discarded.
+                        return truncate_segment(seg, rps->at(1).pos + 12).then([seg, rps, &log] {
+                            return db::commitlog::read_log_file(seg, db::commitlog::descriptor::FILENAME_PREFIX, service::get_local_commitlog_priority(), [rps](db::commitlog::buffer_and_replay_position buf_rp) {
+                                auto&& [buf, rp] = buf_rp;
+                                BOOST_CHECK_EQUAL(rp, rps->at(0));
+                                return make_ready_future<>();
+                            }).then_wrapped([](auto&& f) {
+                                try {
+                                    f.get();
+                                    BOOST_FAIL("Expected exception");
+                                } catch (commitlog::file_truncated_error& e) {
+                                    // ok.
+                                }
+                            });
+                        });
+                    });
+        });
```

Comment:
Why not write the test as coroutine/thread? Less effort later.

Reply:
Test case is modified copy-paste from previous case. Could/should be both coroutine and broken into a helper routine, as there are at least three tests with virtually the same body. Thus keeping it as it is makes it easier to later refactor into helper + cases. And quicker + safer to get test case added now.